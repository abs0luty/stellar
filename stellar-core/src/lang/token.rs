use lasso::Spur;

use crate::lang::location::{Location, Span, Spanned};

use super::ast::{BinaryOperator, BinaryOperatorKind, PrefixOperator, PrefixOperatorKind};

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Keyword {
    With,
    Wait,
    Sequence,
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Punctuation {
    Exclamation,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    LeftParen,
    RightParen,
    Plus,
    Minus,
    PlusEq,
    MinusEq,
    Dot,
    Colon,
    Comma,
}

impl Punctuation {
    pub fn into_binary_operator_kind(&self) -> Option<BinaryOperatorKind> {
        match self {
            Self::Plus => Some(BinaryOperatorKind::Plus),
            Self::Minus => Some(BinaryOperatorKind::Minus),
            _ => None,
        }
    }

    pub fn into_prefix_operator_kind(&self) -> Option<PrefixOperatorKind> {
        match self {
            Self::Exclamation => Some(PrefixOperatorKind::Exclamation),
            _ => None,
        }
    }
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub struct Identifier {
    name: Spur,
    span: Span,
}

impl Identifier {
    pub fn new(name: Spur, span: Span) -> Self {
        Self { name, span }
    }

    pub fn name(&self) -> Spur {
        self.name
    }
}

impl Spanned for Identifier {
    fn span(&self) -> Span {
        self.span
    }
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Token {
    Keyword {
        keyword: Keyword,
        span: Span,
    },
    Identifier(Identifier),
    Punctuation {
        punctuation: Punctuation,
        span: Span,
    },
    Float {
        value: f64,
        span: Span,
    },
    Integer {
        value: i64,
        span: Span,
    },
    Bool {
        value: bool,
        span: Span,
    },
    EOL {
        span: Span,
    },
    EOF {
        /// Location of the last byte in the source file.
        location: Location,
    },
}

impl Token {
    pub fn is_eof(&self) -> bool {
        matches!(self, Self::EOF { .. })
    }

    pub fn is_keyword(&self, keyword: Keyword) -> bool {
        match self {
            Self::Keyword {
                keyword: my_keyword,
                ..
            } => keyword == *my_keyword,
            _ => false,
        }
    }

    pub fn is_punctuation(&self, punctuation: Punctuation) -> bool {
        match self {
            Self::Punctuation {
                punctuation: my_punctuation,
                ..
            } => punctuation == *my_punctuation,
            _ => false,
        }
    }

    pub fn into_binary_operator(&self) -> Option<BinaryOperator> {
        match self {
            Self::Punctuation { punctuation, span } => {
                let kind = punctuation.into_binary_operator_kind()?;

                Some(BinaryOperator { kind, span: *span })
            }
            _ => None,
        }
    }

    pub fn into_prefix_operator(&self) -> Option<PrefixOperator> {
        match self {
            Self::Punctuation { punctuation, span } => {
                let kind = punctuation.into_prefix_operator_kind()?;

                Some(PrefixOperator { kind, span: *span })
            }
            _ => None,
        }
    }

    pub fn is_identifier(&self) -> bool {
        matches!(self, Self::Identifier { .. })
    }
}

impl Spanned for Token {
    fn span(&self) -> Span {
        match self {
            Self::EOF { location } => Span::new(
                *location,
                Location::new(location.line(), location.column() + 1, location.index() + 1),
            ),
            Self::Identifier(Identifier { span, .. })
            | Self::Punctuation { span, .. }
            | Self::Keyword { span, .. }
            | Self::Integer { span, .. }
            | Self::Float { span, .. }
            | Self::Bool { span, .. }
            | Self::EOL { span } => *span,
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct TokenStream(Vec<Token>);

impl TokenStream {
    /// Creates a new empty token stream.
    pub fn new() -> Self {
        Self(Vec::new())
    }

    /// Appends a token to the stream.
    pub fn push(&mut self, token: Token) {
        self.0.push(token);
    }

    /// Returns token with a specified index in the stream. In case index
    /// is out of bounds, EOF token (End Of File) is returned.
    fn get(&self, index: usize) -> Token {
        if index > self.0.len() {
            self.0.last().copied().unwrap_or(Token::EOF {
                location: Location::sof(),
            })
        } else {
            self.0[index]
        }
    }

    /// Returns a cursor over the token stream. See [`TokenStreamCursor`] for more details.
    pub fn into_cursor(self) -> Option<TokenStreamCursor> {
        // Ensure last token is EOF.
        if self.0.last().map_or(true, |maybe_eof| !maybe_eof.is_eof()) {
            return None;
        }

        Some(TokenStreamCursor::new(self))
    }
}

/// A cursor for navigating through a stream of tokens.
///
/// This struct provides functionality to sequentially traverse
/// a [`TokenStream`], allowing you to retrieve tokens one at a time
/// or peek at the upcoming token without advancing the cursor.
/// It tracks the current position in the stream and ensures
/// that an **EOF (End Of File) token is returned when no more
/// tokens are available**.
pub struct TokenStreamCursor {
    stream: TokenStream,
    location: usize,
}

impl TokenStreamCursor {
    fn new(stream: TokenStream) -> Self {
        Self {
            stream,
            location: 0,
        }
    }

    /// Retrieves the next token from the stream, advancing the cursor
    /// to the subsequent position. If no more tokens are available,
    /// an EOF (End Of File) token is returned.
    pub fn next(&mut self) -> Token {
        self.location += 1;

        self.stream.get(self.location - 1)
    }

    /// Provides a glimpse of the next token without advancing the cursor
    /// (compared to [`TokenStreamCursor::next`]). If no more tokens are 
    /// available, an EOF (End Of File) token is returned.
    pub fn peek(&mut self) -> Token {
        self.stream.get(self.location)
    }
}

impl IntoIterator for TokenStream {
    type IntoIter = std::vec::IntoIter<Self::Item>;
    type Item = Token;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}