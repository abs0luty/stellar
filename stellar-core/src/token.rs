use lasso::Spur;

use crate::location::{ByteLocation, Span, Spanned};

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Keyword {
    WithFx,
    WithSynth,
    Wait
}

#[derive(Debug, PartialEq, Clone)]
pub enum Token {
    Keyword {
        keyword: Keyword,
        span: Span,
    },
    Identifier {
        name: Spur,
        span: Span,
    },
    EOF {
        /// Location of the last byte in the source file.
        last_byte_location: ByteLocation,
    },
}

impl Spanned for Token {
    fn span(&self) -> Span {
        match self {
            Self::EOF { last_byte_location} => Span {
                start: *last_byte_location,
                end: last_byte_location.next_byte_location(),
            },
            Self::Identifier { span, .. }
            | Self::Keyword { span, .. } => span.to_owned(),
        }
    }
}
