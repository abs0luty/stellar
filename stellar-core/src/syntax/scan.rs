use crate::{
    match_single_and_two_character_tokens,
    syntax::{
        cursor::Cursor,
        location::{Span, Spanned},
        string_id::StringId,
        token::{Keyword, Operator, Punctuator, Token, TokenStream},
    },
};

use super::token::Identifier;

/// Scans a given Stellar source text and converts into a vector of tokens.
///
/// # Errors
/// In case any obvious syntax errors, which affected the scanning process
/// were found, [`ScannerError`] will be returned.
pub fn scan(source: &str) -> Result<TokenStream, ScanError> {
    let mut cursor = Cursor::new(source);
    let mut stream = TokenStream::new();

    loop {
        let token = scan_next_token(&mut cursor)?;

        if let Token::EndOfFile { .. } = token {
            stream.push(token);

            break;
        }

        stream.push(token);
    }

    Ok(stream)
}

/// Scans the next token in the source text and advances position of the [`Cursor`].
fn scan_next_token(cursor: &mut Cursor) -> Result<Token, ScanError> {
    while let Some(c) = cursor.peek() {
        match c {
            // Skip whitespace (except line breaks).
            c if c.is_whitespace() && c != '\n' => {
                cursor.next();
            }
            // Skip single-line comments starting with '#'.
            '#' => {
                while let Some(c) = cursor.next() {
                    if c == '\n' {
                        break; // Stop skipping at the end of the line.
                    }
                }
            }
            _ => break, // Stop skipping when non-whitespace and non-comment are found.
        }
    }

    let Some(c) = cursor.peek() else {
        return Ok(Token::EndOfFile {
            location: cursor.location(),
        });
    };

    match c {
        '\n' => {
            let location = cursor.location();
            cursor.next();

            Ok(Token::EndOfLine {
                span: Span::new(location, cursor.location()),
            })
        }
        '"' => scan_string(cursor),
        c if c.is_alphabetic() || c == '_' => Ok(scan_name(cursor)),
        c if c.is_numeric() || c == '.' => Ok(scan_number_or_dot(cursor)),
        _ => {
            let start = cursor.location();
            cursor.next();

            match_single_and_two_character_tokens!(
                c,
                cursor,
                start,
                {
                    '{' => Punctuator::LeftBrace,
                    '}' => Punctuator::RightBrace,
                    '[' => Punctuator::LeftBracket,
                    ']' => Punctuator::RightBracket,
                    '(' => Punctuator::LeftParen,
                    ')' => Punctuator::RightParen,
                    ':' => Punctuator::Colon,
                    '.' => Punctuator::Dot,
                    ',' => Punctuator::Comma,
                },
                {
                    '-' => Operator::Minus,
                    '+' => Operator::Plus,
                    '*' => Operator::Star,
                    '/' => Operator::Slash,
                    '=' => Operator::Assign,
                },
                {
                    '-', '=' => Operator::MinusEq,
                    '+', '=' => Operator::PlusEq,
                    '=', '=' => Operator::Eq,
                }
            )
        }
    }
}

/// Scans a next candidate for identifier token in the source text and if
/// its name matches any known keywords returns keyword token.
fn scan_name(cursor: &mut Cursor) -> Token {
    let mut name = String::new();
    let start = cursor.location();

    while let Some(c) = cursor.peek() {
        if !c.is_alphanumeric() && c != '_' {
            break;
        }

        name.push(c);
        cursor.next();
    }

    let span = Span::new(start, cursor.location());

    fn keyword_from_name(name: &str) -> Option<Keyword> {
        match name {
            "with" => Some(Keyword::With),
            "wait" => Some(Keyword::Wait),
            "sequence" => Some(Keyword::Sequence),
            "play" => Some(Keyword::Play),
            "let" => Some(Keyword::Let),
            "load_sample" => Some(Keyword::LoadSample),
            _ => None,
        }
    }

    match name.as_str() {
        "true" => Token::Bool { value: true, span },
        "false" => Token::Bool { value: false, span },
        name => {
            if let Some(keyword) = keyword_from_name(name) {
                Token::Keyword { keyword, span }
            } else {
                Token::Identifier(Identifier::new(StringId::new(name), span))
            }
        }
    }
}

/// Scans a number or a dot (`.`) from the source text.
fn scan_number_or_dot(cursor: &mut Cursor) -> Token {
    let start = cursor.location();
    let mut has_dot = false;

    while let Some(c) = cursor.peek() {
        if c.is_numeric() {
            cursor.next();
        } else if c == '.' && !has_dot {
            has_dot = true;
            cursor.next();
        } else {
            break;
        }
    }

    let end = cursor.location();

    if end.index() - start.index() == 1 && has_dot {
        return Token::Punctuator {
            punctuator: Punctuator::Dot,
            span: Span::new(start, end),
        };
    }

    let lexeme = &cursor.source()[(start.index() as usize)..(end.index() as usize)];

    if has_dot {
        Token::Float {
            value: lexeme.parse::<f64>().unwrap(),
            span: Span::new(start, end),
        }
    } else {
        Token::Integer {
            value: lexeme.parse::<i64>().unwrap(),
            span: Span::new(start, end),
        }
    }
}

/// Scans a string literal.
fn scan_string(cursor: &mut Cursor) -> Result<Token, ScanError> {
    let start = cursor.location();
    cursor.next(); // Consume the opening quote.

    let mut content = String::new();

    while let Some(c) = cursor.next() {
        match c {
            '\\' => {
                // Handle escape sequences.
                if let Some(escaped_char) = cursor.next() {
                    match escaped_char {
                        'n' => content.push('\n'),
                        't' => content.push('\t'),
                        'r' => content.push('\r'),
                        '"' => content.push('"'),
                        '\\' => content.push('\\'),
                        _ => {
                            return Err(ScanError::InvalidEscapeSequence {
                                span: Span::new(start, cursor.location()),
                                character: escaped_char,
                            })
                        }
                    }
                } else {
                    return Err(ScanError::UnterminatedString {
                        span: Span::new(start, cursor.location()),
                    });
                }
            }
            '"' => {
                // End of string.
                let end = cursor.location();
                return Ok(Token::String {
                    value: StringId::new(content),
                    span: Span::new(start, end),
                });
            }
            _ => {
                content.push(c);
            }
        }
    }

    // Unterminated string.
    Err(ScanError::UnterminatedString {
        span: Span::new(start, cursor.location()),
    })
}

#[derive(Debug, PartialEq)]
pub enum ScanError {
    UnexpectedCharacter { character: char, span: Span },
    InvalidEscapeSequence { character: char, span: Span },
    UnterminatedString { span: Span },
}

impl Spanned for ScanError {
    fn span(&self) -> Span {
        match self {
            Self::UnexpectedCharacter { span, .. }
            | Self::InvalidEscapeSequence { span, .. }
            | Self::UnterminatedString { span } => *span,
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use crate::test_scan;

    use super::scan;

    test_scan!(
        (eof, ""),
        (unexpected_char, "!"),
        (punctuation, "("),
        (number_and_dot, "3 3.2."),
        (name, "wait time"),
        (string, r#""\"Hello,\n \t world\"""#),
    );
}
