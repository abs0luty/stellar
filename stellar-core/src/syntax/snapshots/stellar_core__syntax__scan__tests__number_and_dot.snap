---
source: stellar-core/src/syntax/scan.rs
expression: "scan(\"3 3.2.\")"
snapshot_kind: text
---
Ok(
    TokenStream(
        [
            Integer {
                value: 3,
                span: Span {
                    start: Location {
                        line: 1,
                        column: 0,
                        index: 0,
                    },
                    end: Location {
                        line: 1,
                        column: 1,
                        index: 1,
                    },
                },
            },
            Float {
                value: 3.2,
                span: Span {
                    start: Location {
                        line: 1,
                        column: 2,
                        index: 2,
                    },
                    end: Location {
                        line: 1,
                        column: 5,
                        index: 5,
                    },
                },
            },
            Punctuator {
                punctuator: Dot,
                span: Span {
                    start: Location {
                        line: 1,
                        column: 5,
                        index: 5,
                    },
                    end: Location {
                        line: 1,
                        column: 6,
                        index: 6,
                    },
                },
            },
            EndOfFile {
                location: Location {
                    line: 1,
                    column: 6,
                    index: 6,
                },
            },
        ],
    ),
)
